watcher.go

```diff
@@ -25,14 +25,97 @@ type Watcher struct {
 	ctx       context.Context
 }
 
+func DefaultUpdateCallback(e casbin.IEnforcer) func(string) {
+	return func(msg string) {
+		msgStruct := &MSG{}
+
+		err := msgStruct.UnmarshalBinary([]byte(msg))
+		if err != nil {
+			log.Println(err)
+			return
+		}
+
+		var res bool
+		switch msgStruct.Method {
+		case UpdateTypeUpdate, UpdateTypeUpdateForSavePolicy:
+			err = e.LoadPolicy()
+			res = true
+		case UpdateTypeUpdateForAddPolicy:
+			params := msgStruct.Params[0].([]interface{})
+			rule := make([]string, len(params))
+			for i, v := range params {
+				rule[i] = v.(string)
+			}
+			res, err = e.SelfAddPolicy(msgStruct.Sec, msgStruct.Ptype, rule)
+		case UpdateTypeUpdateForAddPolicies:
+			params := msgStruct.Params[0].([]interface{})
+			rules := make([][]string, len(params))
+			for i, v := range params {
+				rule := make([]string, len(v.([]interface{})))
+				for j, vv := range v.([]interface{}) {
+					rule[j] = vv.(string)
+				}
+				rules[i] = rule
+			}
+			res, err = e.SelfAddPolicies(msgStruct.Sec, msgStruct.Ptype, rules)
+		case UpdateTypeUpdateForRemovePolicy:
+			params := msgStruct.Params[0].([]interface{})
+			rule := make([]string, len(params))
+			for i, v := range params {
+				rule[i] = v.(string)
+			}
+			res, err = e.SelfRemovePolicy(msgStruct.Sec, msgStruct.Ptype, rule)
+		case UpdateTypeUpdateForRemoveFilteredPolicy:
+			fieldIndex := int(msgStruct.Params[0].(float64))
+			params := msgStruct.Params[1].([]interface{})
+			fieldValues := make([]string, len(params))
+			for i, v := range params {
+				fieldValues[i] = v.(string)
+			}
+			res, err = e.SelfRemoveFilteredPolicy(msgStruct.Sec, msgStruct.Ptype, fieldIndex, fieldValues...)
+		case UpdateTypeUpdateForRemovePolicies:
+			params := msgStruct.Params[0].([]interface{})
+			rules := make([][]string, len(params))
+			for i, v := range params {
+				rule := make([]string, len(v.([]interface{})))
+				for j, vv := range v.([]interface{}) {
+					rule[j] = vv.(string)
+				}
+				rules[i] = rule
+			}
+			res, err = e.SelfRemovePolicies(msgStruct.Sec, msgStruct.Ptype, rules)
+		default:
+			err = errors.New("unknown update type")
+		}
+		if err != nil {
+			log.Println(err)
+		}
+		if !res {
+			log.Println("callback update policy failed")
+		}
+	}
+}
+
 type MSG struct {
-	Method string
+	Method UpdateType
 	ID     string
 	Sec    string
 	Ptype  string
-	Params interface{}
+	Params []interface{}
```

Comment:
How about
```
type MSG struct {
	Method      UpdateType
	ID          string
	Sec         string
	Ptype       string
	Rule        []string
	Rules       [][]string
	FieldIndex  int
	FieldValues []string
}
```
I think the current structure parsing code is too complicated, but this definition only needs ``json.UnmarshalBinary``

Reply:
yeah, I agree, I will work on it.